tests/unit/test_charm.py

```diff
@@ -612,17 +700,333 @@ def test_given_default_config_with_interfaces_when_network_attachment_definition
                 "access-ip": DEFAULT_ACCESS_IP,
                 "access-gateway-ip": ACCESS_GW_IP,
                 "gnb-subnet": GNB_SUBNET,
-                "core-interface": "core-net",
-                "core-ip": CORE_IP,
+                "core-interface": CORE_INTERFACE_NAME,
+                "core-ip": VALID_CORE_IP,
                 "core-gateway-ip": CORE_GW_IP,
             }
         )
         nads = self.harness.charm._network_attachment_definitions_from_config()
         for nad in nads:
             config = json.loads(nad.spec["config"])
-            self.assertEqual(config["master"], nad.metadata.name)
+            self.assertTrue(ACCESS_INTERFACE_NAME or CORE_INTERFACE_NAME in config["master"])
             self.assertEqual(config["type"], "macvlan")
 
+    def test_given_upf_configured_to_run_in_dpdk_mode_when_network_attachment_definitions_from_config_is_called_then_2_nads_are_returned(  # noqa: E501
+        self,
+    ):
+        self.harness.disable_hooks()
+        self.harness.update_config(key_values={"upf-mode": "dpdk"})
+        nads = self.harness.charm._network_attachment_definitions_from_config()
```

Comment:
I know we do this at other places in the unit tests but I don't think it's a good idea to test this private method specifically. If we think it's worth testing on its own, I'd argue it should be in a class of its own (or be a method outside of this class) but since it's not the case, we are validating our implementation of the code and making our tests more brittle.

Reply:
True, we do it other places, but I'd say it's rather exceptional. In this case (and probably other cases as well), I did it like this to avoid a boatload of mocking and patching. Also in this case, I don't necessarily agree that it makes our tests brittle. These do not care do we build up our network annotations - in whatever way it is done, the tests only check whether the output contains what it should. 
I think the choice we have here is:
1. Mock and patch a lot of stuff, build up the harness, fire events, get the statefulset and validate network annotations, or
2. Call the function that builds network annotation and check the output.

And I actually think that moving the logic for building up the Network annotation outside the charm class makes a lot of sense.

Reply:
It makes our tests brittle in the sense that they can break even though the behavior is maintainted. Specifically, if w remove `_network_attachment_definitions_from_config` but keep the same behavior implemented differently, we find ourselves with 20 broken tests.

Reply:
I agree, but on the other hand if we need to patch most of the landscape just to test the creation of Network annotations, we may end up with 20 broken tests as well if we, for instance, change the lib we use to talk to Kubernetes. 
Personally, if I need to, I'd rather fix 20 trivial tests than 20 tests full of patches.
I'm not saying we should start testing all private methods, but I think there are situations where it is justified.

Reply:
Those libs are public classes and methods, yes we expect to change the tests if those change. Here `_network_attachment_definitions_from_config` is private and purely an implementation detail.

Reply:
You are one tough adversary in this discussion :rofl: :rofl: 
I give up. I wrote new tests, which do not call private methods.